### PR TITLE
inadequate filtering of user-name during installation process

### PR DESCRIPTION
**Pull request by dgrammatiko**
#### Steps to reproduce the issue

Install Joomla! and use in the user-name a "forbidden character like:
< > \ " ' % ; ( ) &

In my case the credentials where:
Admin-User: S9nFA7N7x7tL}jXOJqIx&vQk3jSF
JAdmin-Pass: msevXL0Stl,BV7szpufhMwo10^IV8SRs

In the installation process this user/password combination was accepted, installation went fine, but one could not log-in any more.

After I changed the username in the Database login was possible again.
#### Expected result

I would expect that during installation there would be a warning, that prevented me from choosing an username that renders the Superadministrator Login useless. Like the one that is used if one registers a new user in the backend.
#### Actual result

Installation with unusable username is possible. Therefore creating a not usable SuperAdmin account on installation
#### Testing

Install this as new branch https://github.com/dgt41/joomla-cms/archive/admin_username.zip
try to enter username containing < > \ " ' % ; ( ) &

### CODE REVIEW

**Comment by framontb** [comment]
I changed the type to "<field name="admin_user" type="username" ... " but it doesn't stops me from choosing '&&&&' as admin_user. Should I doing something wrong?

**Comment by andrepereiradasilva** [comment]
same result as @framontb

**Comment by dgt41** [comment]
Once you hit the next button the validation should kick in and remove those characters

**Comment by andrepereiradasilva** [comment]
so no warning for the user that is admin name changed?

**Comment by dgt41** [comment]
@andrepereiradasilva yup, and that's not good for UX point of view

**Comment by brianteeman** [comment]
Seriously guys try to test and think a little before rushing to a PR.

**Comment by framontb** [comment]
This configuration:
`<field name="admin_user" type="text" id="admin_user" class="inputbox validate-username"
			label="INSTL_ADMIN_USER_LABEL"
			required="true"
			filter="username"
		/>`
works for me in the way you say @dgt41 . It removes the "bad characters" and go ahead. The user get advised because the text box change to red... but I think it should not go ahead to second step.
Perhaps could you merge your javascript solution with the rule I did? So the registration don't go to second step without a good username.

**Comment by framontb** [comment]
You need to add `validate="installation.adminuser"`, and I think it doesn't need the ` class="validate-username"` (not sure).

File: installation/form/rule/username.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * @package    Joomla.Installation
+ *
+ * @copyright  Copyright (C) 2005 - 2016 Open Source Matters, Inc. All rights reserved.
+ * @license    GNU General Public License version 2 or later; see LICENSE.txt
+ */
+defined('JPATH_BASE') or die;
+/**
+ * Form Rule class for the admin_user field.
+ *
+ * @since  3.5
```
Comment (andrepereiradasilva) [inline installation/form/rule/username.php]: 3.6

**Comment by framontb** [comment]
This is not working as expected because javascript code clean "wrong" characters in admin_user field, so they never reach the server after next-button is pressed. So the rule in server side never applies. It seems time to make a decision?:
- Better javascript solution (intrusive)
- Better php rule in server side solution (non intrusive)
- Modifying  javascript for non changing the admin_user field but alert.

**Comment by bhavikTailored** [comment]
I have tested this item :red_circle: unsuccessfully on ae26aeb09b82ec451d64f391662b7d111f35cda9<br /><br />I have used same credentials which you provided for login. I have successfully login in administrator.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by gunjanpatel** [comment]
I have tested this item :red_circle: unsuccessfully on ae26aeb09b82ec451d64f391662b7d111f35cda9<br /><br />I have tested with https://github.com/joomla/joomla-cms/archive/staging.zip using username and password given in your test instructions while installing joomla.
Admin-User: `S9nFA7N7x7tL}jXOJqIx&vQk3jSF`
JAdmin-Pass: `msevXL0Stl,BV7szpufhMwo10^IV8SRs`

Using the same username and password I was able to login in backend after installing joomla.

Tried with your https://github.com/dgt41/joomla-cms/archive/admin_username.zip and did the same. No difference with joomla staging branch and your admin_username branch.

I have created two video for the same. 
With Joomla staging: https://youtu.be/YS0eSZRqRDQ
With your branch admin_username:  https://youtu.be/skolh5i_FIc

Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by gunjanpatel** [comment]
Sorry @dgt41 please ignore my previous test. I have logged it successfully now and this test is success for.

Using joomla staging branch when I am installing with following,
Admin-User: `S9nFA7N7x7tL}jXOJqIx&vQk3jSF`
JAdmin-Pass: `msevXL0Stl,BV7szpufhMwo10^IV8SRs`

After installation complete I am not able to login.

When I am using your branch `admin_username` i.e. your patch works successfully and it filters characters which are not valid.

screen shot 2016-07-01 at 08 15 08

Thanks for fixing this issue.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by bhavikTailored** [comment]
I have tested this item :white_check_mark: successfully on ae26aeb09b82ec451d64f391662b7d111f35cda9<br /><br />I am sorry, I have tested again and found it successful. I misunderstood `{` and `&` character. It is counting `{` as valid but `&` as invalid character.

Thanks, it works for me so this patch is tested successfully.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by gunjanpatel** [comment]
One weird thing is that I am choosing username `S9nFA7N7x7tL}jXOJqIx&vQk3jSF` while login. Notice `&` is still in username and I am able to login with username having `&` in it while during installation `&` was already filtered out. 
Still second thought in my mind that this is a valid test or now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by gunjanpatel** [comment]
I have tested this item :white_check_mark: successfully on ae26aeb09b82ec451d64f391662b7d111f35cda9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by gunjanpatel** [comment]
@dgt41 I am not sure why but travis is not happy :smile: <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by zero-24** [comment]
@dgt41 @gunjanpatel 

CS error found by Mr. Travis ;)

```
FILE: ...home/travis/build/joomla/joomla-cms/installation/form/rule/username.php
--------------------------------------------------------------------------------
FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
--------------------------------------------------------------------------------
 29 | ERROR | Expected 1 blank line before member var; 0 found
--------------------------------------------------------------------------------
UPGRADE TO PHP_CODESNIFFER 2.0 TO FIX ERRORS AUTOMATICALLY
--------------------------------------------------------------------------------
```

File: installation/form/rule/username.php [hunk 2]
```diff
@@ -9,22 +9,24 @@
 /**
  * Form Rule class for the admin_user field.
  *
- * @since  3.5
+ * @since  3.7
```
Comment (gunjanpatel) [inline installation/form/rule/username.php]: I am not getting why `3.7` here.

Comment (zero-24) [inline installation/form/rule/username.php]: @gunjanpatel 3.6 is planed to be released on 12.07 and the seccond RC Version is out. There should be only bugfixes go into 3.6.0 stable ;)

https://www.joomla.org/announcements/release-news/5663-joomla-3-6-rc-2-released.html

Comment (gunjanpatel) [inline installation/form/rule/username.php]: Great. Thank you. I am getting the logic now. :smile:

**Comment by gunjanpatel** [comment]
I have tested this item :white_check_mark: successfully on c278ba801d22b5697f19a5184bf92c7a41f7201f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by gunjanpatel** [comment]
Setting RTC as we have two successful tests. 
Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by roland-d** [comment]
Removing the RTC label here because I don't think this PR is complete.

There should be a message to the user the username is invalid instead of just cleaning it without notice. The same way as the password field works.

As for my test, I don't see the username being cleaned. I did see it happen once but not anymore after clicking Next and back a few times. It will only highlight the field as incorrect when you focus out of the field. 

Now I found out something else doing the following:
1. I copy-pasted your username into the field
2. Ignored the error
3. Clicked Next
4. Click back and see the ) is stripped out
5. Add the ) back into the username
6. Click Next
7. Click Back
8. Notice the ) is still there<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10391">issues.joomla.org/joomla-cms/10391</a>.</sub>

**Comment by dgt41** [comment]
@roland-d since there is another PR for ajaxifying this field, is this still relative or should I just close it?

**Comment by roland-d** [comment]
@dgt41 Does the other PR also touch the installation part? If so, I guess we can close this one.

**Comment by dgt41** [comment]
@roland-d I have no clue, also I don't know if the checks are only if the name exists or more specific if the name doesn't contain illegal characters. I guess I'll have to test that...

**Pull request closed without merging**